blockchain/blockchain.go

```diff
@@ -209,3 +233,47 @@ func (b *Blockchain) SanityCheckNewHeight(block *core.Block, stateUpdate *core.S
 
 	return core.VerifyBlockHash(block, b.network)
 }
+
+func (b *Blockchain) StoreL1ChainHeight(height uint64, block *core.Block) error {
+	if err := b.VerifyL1ChainHeight(height); err != nil {
+		return err
+	}
+	return b.database.Update(func(txn db.Transaction) error {
+		if err := putBlock(txn, block); err != nil {
+			return err
+		}
+		heightBin := make([]byte, lenOfBlockNumberBytes)
+		binary.BigEndian.PutUint64(heightBin, height)
+		return txn.Set(db.L1ChainHeight.Key(), heightBin)
+	})
+}
+
+func (b *Blockchain) VerifyL1ChainHeight(height uint64) error {
+	txn := b.database.NewTransaction(false)
+	defer txn.Discard()
+	return b.verifyL1ChainHeight(txn, height)
+}
+
+func (b *Blockchain) verifyL1ChainHeight(txn db.Transaction, height uint64) error {
+	l2Height, err := b.height(txn)
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return err
+	}
+	if height > l2Height {
+		return ErrInvalidL1Verification{
+			"l1 height cannot be greater than l2 height",
+		}
+	}
+
+	l1Height, err := b.l1ChainHeight(txn)
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return err
+	}
+	if l1Height+1 != height && l1Height != height {
+		return ErrInvalidL1Verification{
```

Comment:
we shouldnt need to verify blocks one by one from L1. That would require an ethereum archive node. When we implement the actual L1 verification, we should only be listening to new verification events.

For example; if we detect a block on height `N` is verified, if the `N`th block in our storage matches the verified block's hash, we now know just that all previous blocks were L1 verified as well.

Reply:
Yeap, that is another edge case we need to handle